### ISSUE TITLE
xDS: v1.66.2 and above break most xDS client gRPC requests

### ISSUE THREAD

**Issue by marcel808** [grpc/grpc-go#7691]
### What version of gRPC are you using? 
v1.65.0 works, 1.66.2/1.67.0 cause most gRPC xDS based requests to fail with error
```
rpc error: code = Unavailable desc = xds: error received from xDS stream: EOF
```

### What version of Go are you using (`go version`)?
1.22.6

### What operating system (Linux, Windows, …) and version?
Linux (Google GKE)

### What did you do?
We have go service pods that call out to other services using the istio agent (inject.istio.io/templates:a grpc-agent), prefix the service urls with "xds:///"  and import _ "google.golang.org/grpc/xds". 

istiod-1-22-4

### What did you expect to see?
Succesfull gRPC requests load balanced using xDS

### What did you see instead?
90% gRPC failure rate with error: rpc error: code = Unavailable desc = xds: error received from xDS stream: EOF

**Comment by easwars** [comment]
Would you be able to provide some logs for us? With the following env vars set: `GRPC_GO_LOG_VERBOSITY_LEVEL=99 GRPC_GO_LOG_SEVERITY_LEVEL=info`.

Thanks.

**Comment by easwars** [comment]
The error mentioned in the issue description happens when connection to the management server fails. https://github.com/grpc/grpc-go/blob/ca4865d6dd6f3d8b77f1943ccfd6c9e78223912d/xds/internal/xdsclient/authority.go#L462C1-L463C1

**Comment by marcel808** [comment]
I enabled grpc-go version 1.67.2 and here are the logs leading up to the error:
```
2024/10/03 16:01:45 INFO: [core] [Channel #26 SubChannel #27]Subchannel created
2024/10/03 16:01:45 INFO: [core] [Channel #26]Channel Connectivity change to CONNECTING
2024/10/03 16:01:45 INFO: [core] [Channel #26]Channel exiting idle mode
2024/10/03 16:01:45 INFO: [xds] [xds-client 0xc000bd8190] [unix:///etc/istio/proxy/XDS] Created transport to server "unix:///etc/istio/proxy/XDS"
2024/10/03 16:01:45 INFO: [xds] [xds-client 0xc000bd8190] [unix:///etc/istio/proxy/XDS] New watch for type "ListenerResource", resource name "<redacted>.svc.cluster.local:50051"
2024/10/03 16:01:45 INFO: [xds] [xds-client 0xc000bd8190] [unix:///etc/istio/proxy/XDS] First watch for type "ListenerResource", resource name "<redacted>.svc.cluster.local:50051"
2024/10/03 16:01:45 INFO: [core] [Channel #24]Channel exiting idle mode
2024/10/03 16:01:45 INFO: [core] [Channel #26 SubChannel #27]Subchannel Connectivity change to CONNECTING
2024/10/03 16:01:45 INFO: [core] [Channel #26 SubChannel #27]Subchannel picks a new address "/etc/istio/proxy/XDS" to connect
2024/10/03 16:01:45 INFO: [core] original dial target is: "xds:///<redacted>.svc.cluster.local:50052"
2024/10/03 16:01:45 INFO: [core] [Channel #28]Channel created
2024/10/03 16:01:45 INFO: [core] [Channel #28]parsed dial target is: resolver.Target{URL:url.URL{Scheme:"xds", Opaque:"", User:(*url.Userinfo)(nil), Host:"", Path:"/<redacted>.svc.cluster.local:50052", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}}
2024/10/03 16:01:45 INFO: [core] [Channel #28]Channel authority set to "<redacted>.svc.cluster.local:50052"
2024/10/03 16:01:45 INFO: [pick-first-lb] [pick-first-lb 0xc000c1e240] Received SubConn state update: 0xc000c1e2a0, {ConnectivityState:CONNECTING ConnectionError:<nil> connectedAddress:{Addr: ServerName: Attributes:<nil> BalancerAttributes:<nil> Metadata:<nil>}}
2024/10/03 16:01:45 INFO: [core] connect called on addrConn in non-idle state (CONNECTING); ignoring.
2024/10/03 16:01:45 INFO: [xds] [xds-resolver 0xc000a74fc0] Creating resolver for target: xds:///<redacted>.svc.cluster.local:50052
2024/10/03 16:01:45 INFO: [xds] [xds-bootstrap] Using bootstrap file with name "/etc/istio/proxy/grpc-bootstrap.json" from GRPC_XDS_BOOTSTRAP environment variable
2024/10/03 16:01:45 INFO: [core] [Channel #26 SubChannel #27]Subchannel Connectivity change to READY
2024/10/03 16:01:45 INFO: [pick-first-lb] [pick-first-lb 0xc000c1e240] Received SubConn state update: 0xc000c1e2a0, {ConnectivityState:READY ConnectionError:<nil> connectedAddress:{Addr:/etc/istio/proxy/XDS ServerName:localhost Attributes:0xc00021a7a8 BalancerAttributes:<nil> Metadata:<nil>}}
2024/10/03 16:01:45 INFO: [core] [Channel #26]Channel Connectivity change to READY
2024/10/03 16:01:45 INFO: [xds] [xds-client 0xc000bd8190] [unix:///etc/istio/proxy/XDS] ADS stream created
2024/10/03 16:01:45 WARNING: [xds] [xds-client 0xc0001d7b80] [unix:///etc/istio/proxy/XDS] ADS stream closed: EOF
2024/10/03 16:01:45 INFO: [xds] [xds-resolver 0xc00072b440] Received error for Listener resource "mtbe-service.mtbe.svc.cluster.local:8666": xds: error received from xDS stream: EOF
2024/10/03 16:01:45 WARNING: [core] [Channel #1]ccResolverWrapper: reporting error to cc: xds: error received from xDS stream: EOF
```

**Comment by marcel808** [comment]
vs 1.65.0:
```
2024/10/03 15:21:58 INFO: [xds] [xds-client 0xc0007fee10] [unix:///etc/istio/proxy/XDS] ADS stream created
2024/10/03 15:21:58 INFO: [xds] [xds-client 0xc0007fee10] [unix:///etc/istio/proxy/XDS] ADS request sent: {
"node":  {
"id":  "sidecar~<redacted>.cluster.local",

```

**Comment by easwars** [comment]
Thanks for the logs.

```
2024/10/03 16:01:45 WARNING: [xds] [xds-client 0xc0001d7b80] [unix:///etc/istio/proxy/XDS] ADS stream closed: EOF
```

The above line seem to indicate to me that the server is closing the stream for whatever reason. The above is logged when the xDS client runs into an error when attempting to read from the ADS stream.

**Comment by easwars** [comment]
Does your server logs contain anything useful?

**Comment by marcel808** [comment]
Could this be related?
```
{"level":"warn","time":"2024-10-03T16:01:45.814940Z","scope":"ads","msg":"ADS: \"dd.dd.15.80:37586\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:45.820678Z","scope":"ads","msg":"ADS: \"dd.dd.15.79:57722\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:46.037482Z","scope":"ads","msg":"ADS: \"dd.dd.0.133:35194\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:46.426842Z","scope":"ads","msg":"ADS: \"dd.dd.4.251:50008\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:46.823614Z","scope":"ads","msg":"ADS: \"dd.dd.15.80:37612\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:47.610846Z","scope":"ads","msg":"ADS: \"dd.dd.8.49:60104\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:47.680695Z","scope":"ads","msg":"ADS: \"dd.dd.17.87:40414\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:47.894435Z","scope":"ads","msg":"ADS: \"dd.dd.0.133:35212\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:48.303335Z","scope":"ads","msg":"ADS: \"dd.dd.17.86:56104\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:48.446480Z","scope":"ads","msg":"ADS: \"dd.dd.9.82:22297\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:48.598658Z","scope":"ads","msg":"ADS: \"dd.dd.15.79:55362\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:48.676814Z","scope":"ads","msg":"ADS: \"dd.dd.17.86:56120\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:49.729190Z","scope":"ads","msg":"ADS: \"dd.dd.17.86:56146\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:50.035842Z","scope":"ads","msg":"ADS: \"dd.dd.8.49:60114\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:50.945518Z","scope":"ads","msg":"ADS: \"dd.dd.0.133:36854\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:52.271491Z","scope":"ads","msg":"ADS: \"dd.dd.17.86:56196\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:52.466689Z","scope":"ads","msg":"ADS: \"dd.dd.9.82:35128\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:54.651522Z","scope":"ads","msg":"ADS: \"dd.dd.15.79:55434\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:01:59.183774Z","scope":"ads","msg":"ADS: \"dd.dd.17.86:41806\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:02:01.358069Z","scope":"ads","msg":"ADS: \"dd.dd.17.86:41826\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:02:01.572871Z","scope":"ads","msg":"ADS: \"dd.dd.17.87:33432\" exceeded rate limit: context canceled"}
{"level":"warn","time":"2024-10-03T16:02:01.891733Z","scope":"ads","msg":"ADS: \"dd.dd.0.133:56666\" exceeded rate limit: context canceled"}

```

**Comment by easwars** [comment]
Interesting. Is this from your server?

But from the client logs, it seems like the client is not even sending one ADS request, right?

**Comment by marcel808** [comment]
> Interesting. Is this from your server?
Yes, this is from kubectl logs -n istio-system istiod-....
> 
> But from the client logs, it seems like the client is not even sending one ADS request, right?

Seems like it since like it since the working version shows ADS request sent. I'll ask our infra structure team whether there are certain rate limits in play.
Note sure whether the rate limits are the cause of the EOF errors but they occur around the same time.
Any idea how the changes post 1.65 could have affected this?

**Comment by easwars** [comment]
> Any idea how the changes post 1.65 could have affected this?

There were a few changes that could be of interest:
- https://github.com/grpc/grpc-go/pull/7458: This change causes the xDS client to block further reads on the ADS stream until the previous update is completely processed locally
- https://github.com/grpc/grpc-go/pull/7347: This change causes gRPC to create one xDS client per gRPC dial target. Earlier, there used to be single xDS client per gRPC binary. Now, if your binary makes multiple gRPC channels to different target URIs, you would have multiple xDS clients, one each for each target URI
  - This could cause more connections and more ADS streams on your server

**Comment by marcel808** [comment]
Thanks for the reference, I'll see also if the metric can located here that shows the ADS request rate to confirm there's a link to this issue.

**Comment by marcel808** [comment]
Adjusted the title as there might be some pods able to make requests. I will gather more information on the gRPC service and client pod counts. The istiod YAML does not show any limits that might be involved in the xDS requests; I do see the istio operator setting variables like PILOT_MAX_REQUESTS_PER_SECOND=100 and PILOT_PUSH_THROTTLE=100 which I think affect the agent in the client pod but will need to read up on this.

**Comment by marcel808** [comment]
background
==========
QA GKE cluster with limited number of pods (to keep cost down).  There are 3 pods that were upgraded to with different service dependencies:

- a: 4 deps
- b: 2
- c: 2

e.g. the grpc-go change to create separate xDS clients for each service dependencies brings with it additional ADS request load on the istiod side.

client side observations
========================
- ADS requests (as logged by grpc-go) from the upgraded pods increased on a per minute basis by 10x. The h/l/t prefixes identify different services while the 15:00 hour counts are v1.65.0 vs 16:00 are v1.67.1.  Note that the logged request counts don't seem to add up to the 100 PILOT_MAX_REQUESTS_PER_SECOND (e.g. 6k/minute).

- Nearly all client gRPC requests fail when the upgrade to grpc-go v1.67.1 is deployed (~ 12 ET - 16:00 UTC). See graph. 
- The services never seem to get into a stable state, e.g. the only way to recover is to roll back the deploy back to v1.65.0

![image](https://github.com/user-attachments/assets/1a14a1a9-e65f-47c0-86d8-df4bd84e62bc)

istiod obverations
==============
- Rate limit errors happen also before switching the service to v1.67.1 but do increase. The per minute rate is lower than the gRPC xdDS stream EOF error rate on the client side. 

![image](https://github.com/user-attachments/assets/44e06c61-9be6-4f91-8776-179bfe5a8bc4)

- The (3) istiod pods did not show signs of being overloaded (cpu/memory wise) during the the time of v1.67.1  change.

So in summary, it doesn't seem like the istiod side logged ADS rate limit errors account for a near 100% failure rate of gRPC calls for xDS clients. Also, if there's a temporary case of the 100/s rate limit for ADS requests being exceeded (potentially due to the increased number of xDS clients per > 1.65.0 changes) I would expect the services eventually being able to recover (and get ADS responses). In this QA scenario they don't seem to be able to.

**Comment by easwars** [comment]
Thanks for the detailed report. 

So, if the stream errors seen by gRPC are not because of rate limit errors, does the istiod logs have anything useful about why it is closing streams? gRPC is seeing EOF on stream reads, so there is not anything else useful from gRPC logs about why the stream failed.

Is there anyway you can provide us with a repro that we can use on our side? I personally don't have much experience with configuring and running istiod.

**Comment by marcel808** [comment]
The istiod logs are not showing much besides rate limit warnings (no errors). What is interesting though what's not logged: I see :"ADS: new delta connection for node:..." for services that run a plain istio-proxy side car but not for the service that has problems (it uses the istio-proxy in agent mode). 
The agent logs I will pull today to see the difference in output between the 1.65.0 usage and 1.67 since that sits in between the service and the istiod.

Will make attempt to reproduce the problem with some dummy services/clients and minikube.

**Comment by easwars** [comment]
> ADS: new delta connection for node 

That is interesting. gRPC does not support the delta variant of the xDS protocol. gRPC currently only supports the SotW variant. So, maybe the delta connection is originating from something else?

**Comment by marcel808** [comment]
I compared the istio-proxy logs for agent vs "normal" mode and found this difference:

```
agent:
{"level":"info","time":"2024-10-10T12:47:42.542125Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":81}

"normal (and logged by istiod)":
{"level":"info","time":"2024-10-10T09:03:02.891659Z","scope":"xdsproxy","msg":"connected to delta upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":129}
```

Going to try running the service with a plain proxy and 1.67.2 and see what happens; at least it would narrow the problem down.

**Comment by marcel808** [comment]
Switching to non-agent opened another can of worms so sticking with agent mode for now (this is also what we prefer as these ae very high volume services that get bogged down by the overhead of the proxy using up more CPU than the service container itself processing every byte going in and out)

The agent logs show this:
```
{"level":"info","time":"2024-10-11T14:47:00.012793Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":54}
{"level":"info","time":"2024-10-11T14:47:00.012957Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":53}
{"level":"warn","time":"2024-10-11T14:47:00.279688Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"warn","time":"2024-10-11T14:47:00.280400Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"info","time":"2024-10-11T14:47:00.286153Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":55}
{"level":"info","time":"2024-10-11T14:47:00.287524Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":56}
{"level":"warn","time":"2024-10-11T14:47:05.700289Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"warn","time":"2024-10-11T14:47:05.700923Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
```
https://github.com/istio/istio/issues/37152 has a comment on this:
```
I think the 'overlapping' message in Istiod is the key - if I remember correctly, it happens if 2 connections with the same node ID connect to Istiod - and istio believes it failed to detect the RST/FIN on the first connection ( a client should have a single active XDS connection), so it closes it. And it gets into a loop.

Please make sure you didn't cut&paste a bootstrap file in 2 places. Normally the generated node id has a unique name, based on pod.
```

This sounds connected to the change to make different xDS clients per service name connected to?

**Comment by marcel808** [comment]
One issue that might be happening is that istiod is not expecting multiple xDS connections with the same node is, in our case it's set as:
```
            - name: ISTIO_BOOTSTRAP_OVERRIDE
              value: "/etc/istio/custom-bootstrap/custom_bootstrap.json"

                  NODE_ID="sidecar~${INSTANCE_IP}~${POD_NAME}.${POD_NAMESPACE}~cluster.local"
```

**Comment by marcel808** [comment]
In short, the istio-agent seems to be closing the xDS clients beyond the first due to the single node_id used. https://github.com/grpc/grpc-go/pull/7347 requires quite some understanding of both grpc-go and istio xDS to understand have the named clients would work, when the agent doesn't seem to be aware of the "name" part of the xDS clients.

This is the typical startup behavior of the istio agent where the first xDS connection works but all subsequent attempts get closed, hence of the gRPC client call xds EOF failures:
```
{"level":"info","time":"2024-10-11T17:57:26.160505Z","scope":"xdsproxy","msg":"Initializing with upstream address \"istiod-1-22-4.istio-system.svc:15012\" and cluster \"Kubernetes\""}
{"level":"info","time":"2024-10-11T17:57:26.427983Z","scope":"cache","msg":"generated new workload certificate","resourceName":"default","latency":267051261,"ttl":86399572022083}
{"level":"info","time":"2024-10-11T17:57:26.428066Z","scope":"cache","msg":"Root cert has changed, start rotating root cert"}
{"level":"info","time":"2024-10-11T17:57:26.428100Z","scope":"cache","msg":"returned workload certificate from cache","ttl":86399571901173}
{"level":"info","time":"2024-10-11T17:57:26.428366Z","scope":"cache","msg":"returned workload trust anchor from cache","ttl":86399571634504}
{"level":"info","time":"2024-10-11T17:57:26.428802Z","scope":"cache","msg":"returned workload trust anchor from cache","ttl":86399571198514}
{"level":"info","time":"2024-10-11T17:57:33.536595Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":1}
{"level":"warn","time":"2024-10-11T17:57:33.808855Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"warn","time":"2024-10-11T17:57:33.809498Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"info","time":"2024-10-11T17:57:33.815859Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":2}
{"level":"info","time":"2024-10-11T17:57:33.816460Z","scope":"xdsproxy","msg":"connected to upstream XDS server: istiod-1-22-4.istio-system.svc:15012","id":3}
{"level":"warn","time":"2024-10-11T17:57:34.817348Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}
{"level":"warn","time":"2024-10-11T17:57:34.818000Z","scope":"xdsproxy","msg":"registered overlapping stream; closing previous"}

```

**Comment by easwars** [comment]
> https://github.com/grpc/grpc-go/pull/7347 requires quite some understanding of both grpc-go and istio xDS as to know how multiple named clients would work, when the agent doesn't seem to be aware of the "name" part of the xDS clients.

The "name" is completely local to grpc and is not part of the xDS protocol. So, the name will not be communicated to the xDS peer.

Do you know why the istio-agent closes xDS clients beyond the first one? Can it be configured to allow multiple xDS clients with the same node ID? All xDS clients from the same gRPC binary should be using the same node ID.

**Comment by marcel808** [comment]
Note sure; It might just an assumption that a pod should only open one xDS connecton during its lifetime. Opening up a ticket on the istio project might clarify or provide some recommendations, linking back to the this ticket.

**Comment by easwars** [comment]
Thanks for filing the issue with istio. We do feel that it is a bug on the istio side to close xDS connections from gRPC other than the first one. Let's see what they say.

**Comment by bozaro** [comment]
This issue is show-stopper for go-grpc upgrade. Are there some workaround?

**Comment by easwars** [comment]
[gRFC A97](https://github.com/grpc/proposal/blob/master/A97-xds-jwt-call-creds.md) is a proposal to work around this issue and has been implemented in Go with the [following PRs](https://github.com/grpc/grpc-go/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Amerged+%22A97%22+). This should be part of the upcoming v1.77.0 release.

@marcel808 : Please let us know if this works for you.

**Comment by marcel808** [comment]
@easwars thanks for the fix, will let you know.

**Comment by marcel808** [comment]
@easwars A97 does not seem to address obtaining the channel credentials and JWT tokens to access the xDS service. For Google GKE with plain istio, where would one find the information to obtain these and store them in the container files (and update them before expiry)? We do not have the bandwidth to reverse engineer the istio agent to see how it gets its credentials and replicate it with our own scripts/code.

**Comment by easwars** [comment]
@marcel808 : Are you still using the istio agent to proxy the xDS connection? From https://github.com/istio/istio/issues/53532, it looks like the general plan was to not use the istio agent for proxying. Is that something you are considering?

@dimpavloff: Would you be able to share how you got rid of the problem on your deployments post A97 implementation? Thanks.

**Comment by marcel808** [comment]
@easwars Yes, we've pinned our gRPC go clients to v1.65.1 since last year and using the istio agent to make use of xDS load distribution. Not using the agent is not a problem. I might have misread the A97 doc: can we eventually upgrade to v1.77.0 and xDS gRPC requests would just work (e.g. authenticate with the istio xDS service without problems)?

**Comment by easwars** [comment]
> can we eventually upgrade to v1.77.0 and xDS gRPC requests would just work (e.g. authenticate with the istio xDS service without problems)?

You need to set up a JWT token on the local file system and have a bootstrap file that has the newly added `call_credentials` field that points to this file on the local file system. I don't know enough about istio and your deployment to help you with how to do this though.

**Comment by marcel808** [comment]
> > can we eventually upgrade to v1.77.0 and xDS gRPC requests would just work (e.g. authenticate with the istio xDS service without problems)?
> 
> You need to set up a JWT token on the local file system and have a bootstrap file that has the newly added `call_credentials` field that points to this file on the local file system. I don't know enough about istio and your deployment to help you with how to do this though.

This is for Google GKE with a standard (I think) istio install. I'll see whether our own infrastructure team can help with this or use our Google support.

**Comment by dimpavloff** [comment]
> @dimpavloff: Would you be able to share how you got rid of the problem on your deployments post A97 implementation? 

Hey; I was planning on figuring out what bootstrap file to generate for the Istio (e.g. by adopting what the pilot-agent does and adjusting the server URI and anything else along the way as per this comment https://github.com/istio/istio/issues/53532#issuecomment-2610971115 ) 

However, I haven't had the chance to do this yet but should have some time in a couple of weeks. In the meantime, I'm also keeping an eye on https://github.com/istio/istio/pull/58056

**Comment by easwars** [comment]
@kyessenov

Can you help with this or can you point us in the right direction to get help from the appropriate person. Thank you.

**Comment by kyessenov** [comment]
Sorry, Istio OSS decision was to keep it simple and only support a single xDS connection for envoy and ztunnel. I am not really on any governance committee there and there are already issues and PRs open in Istio.

**Comment by laz-canva** [comment]
> This is for Google GKE with a standard (I think) istio install. I'll see whether our own infrastructure team can help with this or use our Google support.

On GCP, Traffic Director (now Cloud Service Mesh) is worth considering instead of Istio. It's designed for the scale where direct gRPC connections make sense. Istio seems more focused on a full proxy mesh where one cannot control client and server code.

**Comment by kyessenov** [comment]
@laz-canva Yes, I think these are users of gRPC mesh. The issue at hand is proxy-control plane communication, and there's some problem with authentication setup (using k8s JWT to authenticate to gRPC servers) that necessitated using an OSS istio agent. My take is that gRPC needs to solve it directly without a sidecar since sidecars generally negate the benefit of gRPC as the mesh solution.

**Comment by laz-canva** [comment]
> My take is that gRPC needs to solve it directly without a sidecar since sidecars generally negate the benefit of gRPC as the mesh solution.

Yes, no sidecars is the ideal state. gRPC can do this for the data plane, but in the control plane the gRPC client has to interoperate with Istio's sidecar focused xDS implementation. This integration work falls between projects, and progress is slow.

**Comment by kyessenov** [comment]
@laz-canva Istio ultimately is another k8s app using k8s JWT as the identity source, and authentication using k8s JWT is fairly standard. I think it would be nice to precisely define the gaps in gRPC and work on solving them. @markdroth made some progress, and I thought it might be all of it, but it looks it is not.

**Comment by easwars** [comment]
The gRPC team does not have the cycles to prioritize this work at this point in time. We plan to revisit this issue in Q2 of 2026. Thank you for your patience.
